pycbc/events/coinc.py

```diff
@@ -370,7 +370,97 @@ def cluster_coincs(stat, time1, time2, timeslide_id, slide, window, argmax=numpy
 
     span = (time.max() - time.min()) + window * 10
     time = time + span * tslide
+    cidx = cluster_given_mean_times(stat, time, window, argmax)
+    return cidx
+
+def cluster_coincs_multiifo(stat, time_coincs, timeslide_id, slide, window, argmax=numpy.argmax):
+    """Cluster coincident events for each timeslide separately, across
+    templates, based on the ranking statistic
+
+    Parameters
+    ----------
+    stat: numpy.ndarray
+        vector of ranking values to maximize
+    time_coincs: tuple of numpy.ndarrays
+        tuple containing time vectors each ifo
+    timeslide_id: numpy.ndarray
+        vector that determines the timeslide offset
+    slide: float
+        length of the timeslides offset interval
+    window: float
+        legth to cluster over
+
+    Returns
+    -------
+    cindex: numpy.ndarray
+        The set of indices corresponding to the surviving coincidences
+    """
+    time_coinc_zip = zip(*time_coincs)
+    if len(time_coinc_zip) == 0:
+        logging.info('No coincident triggers.')
+        return numpy.array([])
+
+    time = []
+    #find the mean of each time coincidences, providing that the IFO is on
+    for tc in time_coinc_zip:
+        time.append(mean_if_greater_than_zero(tc))
+
+    num_ifos = [len([t for t in tc if t > 0]) for tc in time_coinc_zip]
+
+    if numpy.isfinite(slide):
+        time = time + ((num_ifos - numpy.ones_like(num_ifos))*timeslide_id * slide)/num_ifos
+
+    tslide = timeslide_id.astype(numpy.float128)
+    time = time.astype(numpy.float128)
+
+    span = (time.max() - time.min()) + window * 10
+    time = time + span * tslide
+    cidx = cluster_given_mean_times(stat, time, window, argmax)
+    return cidx
+
+def mean_if_greater_than_zero(vector):
+    """ Calculate mean of a vector, but ignore it if the value if it is less than zero
+    This is used when the timestamps are marked as -1 when a particular
+    IFO is not included in the particular coincidence.
+
+    Parameters
+    ----------
+    vector: numpy.ndarray
+        vector to be mean averaged
+
+    Returns
+    -------
+    mean: float
+        The mean of the values in the original vector which are
+        greater than zero
+    """
+    num_above_zero = len([val for val in vector if val > 0])
+    mean = sum([val / num_above_zero if val > 0 else 0 for val in vector])
+    return mean
```

Comment:
to avoid having to calculate num above zero again elsewhere, return both num above zero and mean?
also prefer to divide by num above zero outside the loop. 
can't you just do the second loop like the first - ie `val if val > 0 for val in vector` ?

Reply:
Reorganised to make it so that it only does the list comprehension once - now returns both values